### PR TITLE
[JENKINS-72111] Allow `Lifecycle` to load implementations from plugins

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-72111](https://issues.jenkins.io/browse/JENKINS-72111) and https://github.com/jenkinsci/jenkins/pull/8555#discussion_r1349509992. Unlike the original PR, this requires no API change or plugin initializer, just the existing system property—which had long used `UberClassLoader` but did not work for a class defined in a plugin.

### Testing done

The functional test added here, as well as validation that this works for a `Lifecycle` implementation in a plugin in CloudBees CI when the system property is set. (Without this patch, setting the system property to the plugin-defined class causes a boot failure, since the class is not loadable when `Lifecycle.get` is first called early during startup.)

### Proposed changelog entries

- Allow plugins to define a custom `Lifecycle`.

### Desired reviewers

@Vlatombe @jtnord

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: test/src/test/resources/plugins/custom-lifecycle.hpi
Comment (jglick) [inline test/src/test/resources/plugins/custom-lifecycle.hpi:1]: I retained sources in the plugin archive, but for the convenience of reviewers:

```java
package test.custom_lifecycle;
import hudson.lifecycle.Lifecycle;
public final class CustomLifecycle extends Lifecycle {
    public int count;
    @Override
    public void restart() {
        count++;
    }
}
```

**Review by Vlatombe — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Review by jtnord — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
